rust/src/csv/reader.rs

```diff
@@ -52,6 +53,127 @@ use crate::record_batch::RecordBatch;
+                        // possibilities.push(2);
+                    } else if DECIMAL_RE.is_match(v) {
+                        possibilities.push(3);
+                    } else if INTEGER_RE.is_match(v) {
+                        possibilities.push(4);
+                    } else {
+                        // we have a string, so the schema becomes a string
+                        return Field::new(&field_name, DataType::Utf8, false)
+                    }
+                }
+            },
+            None => {
+                // do nothing, we'll return an optional String
+                possibilities.push(0);
+            }
+        }
+    }
+    possibilities.sort();
+    possibilities.dedup();
+    if possibilities.len() == 1 {
+        match possibilities[0] {
+            0 => { return Field::new(&field_name, DataType::Utf8, has_nulls) },
+            1 => { return Field::new(&field_name, DataType::Boolean, has_nulls) },
+            2 => { return Field::new(&field_name, DataType::Int64, has_nulls) },
+            3 => { return Field::new(&field_name, DataType::Float64, has_nulls) },
+            4 => { return Field::new(&field_name, DataType::Int32, has_nulls) },
+            _ => { return Field::new(&field_name, DataType::Utf8, has_nulls) },
+        }
+    } else if possibilities.len() > 2 {
+        return Field::new(&field_name, DataType::Utf8, has_nulls)
+    } else {
+        if possibilities.contains(&1) || possibilities.contains(&2) {
+            // boolean or timestamp is incompatible with other types, return string
+            return Field::new(&field_name, DataType::Utf8, has_nulls)
+        } else if possibilities.contains(&3) && possibilities.contains(&4) {
+            // we have an integer and double, fall down to double
+            return Field::new(&field_name, DataType::Float64, has_nulls)
+        } else {
+            return Field::new(&field_name, DataType::Utf8, has_nulls)
+        }
+    }
+}
+
+/// Infer the schema of a csv file
+fn infer_schema(mut file: File, delimiter: u8) -> Schema {
+
+    file.seek(SeekFrom::Start(0)).unwrap();
+
+    let mut csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // get header names
+    let headers = &csv_reader.headers().unwrap();
+
+    let my_headers: Vec<&str> = headers.clone().iter().collect();
+    let header_length = my_headers.len();
+    let mut my_vals: Vec<Vec<Option<String>>> = vec![vec![]; header_length];
+
+    // return file seek back to start
+    file.seek(SeekFrom::Start(0)).unwrap();
+
+    let csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // take at most 100 rows for schema inference
+    // [TODO] it's often better to read the whole file and infer its schema
+    for result in csv_reader.into_records().take(100) {
```

Comment:
It would be nice to give user the ability to pass in number of records to read as an Optional where the default is to read all rows